scipy/optimize/_zeros_py.py

```diff
@@ -1393,3 +1393,328 @@ def toms748(f, a, b, args=(), k=1,
+
+    Notes
+    -----
+    The order of convergence of the Muller's method is approximately 1.84.
+    This is slightly faster than secant method, but slower than Newton's
+    method which is quadratic.
+
+    When `muller` is used with arrays, it is best suited for the following
+    types of problems:
+
+    * The initial guesses are all relatively the same distance from
+      the roots.
+    * The size of the initial guesses is larger than O(100) elements.
+      Otherwise, a naive loop may perform as well or better than a vector.
+      You should test this for your usecase, as there's significant overhead
+      to using arrays.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy.optimize import muller
+
+    >>> def f(x):
+    ...     return (x**2 + 1)  # roots at x = 1j and x = -1j
+
+    >>> root = muller(f, 0, 1, 2 + 1j)
+    >>> root
+    -1.2168152957711588e-17+1j
+
+    When we want to find roots for a set of related starting values and/or
+    function parameters, we can provide both of those as an array of inputs:
+
+    >>> f2 = lambda x, a: x**3 + a
+    >>> a0 = np.linspace(0, 100, 1001)
+    >>> x0 = np.sqrt(5) * np.linspace(1, 10, 1001)
+    >>> x1 = np.sqrt(3) * np.linspace(1, 10, 1001) * 1j
+    >>> x2 = -np.sqrt(2) * np.linspace(1, 10, 1001)
+
+    >>> vec_res, converged, failed = muller(f2, x0, x1, x2, args=(a0, ), 
+    ...             maxiter=200, full_output=True)
+
+    The above is the equivalent of solving for each value in ``(x, a)``
+    separately in a for-loop, just faster:
+
+    >>> loop_res = [muller(f2, x_0, x_1, x_2, args=(a_0,), maxiter=200)
+    ...             for x_0, x_1, x_2, a_0 in zip(x0, x1, x2, a0)]
+    >>> np.allclose(vec_res, loop_res)
+    True
+
+    Note that in general some values in `vec_res` may not converge, or may 
+    fail as nan.nan due to a poor combination of starting points. These 
+    values are kept to ensure that the array size of `vec_res` is always 
+    identical to sizes of initial guesses.
+
+    In case where size compatibility of `vec_res` is irrelevant (e.g. when
+    searching a `np.linspace` of possible starting points to find different
+    roots), you can easily remove failed and non-converged results:
+
+    >>> vec_res[converged | ~failed]
+
+    """
+
+    if tol <= 0:
+        raise ValueError(f"tol too small ({tol:g} <= 0)")
+    maxiter = operator.index(maxiter)
+    if maxiter < 1:
+        raise ValueError("maxiter must be greater than 0")
+    if np.size(x0) > 1:
+        if np.size(x0) == np.size(x1) == np.size(x2):
+            return _array_muller(f, x0, x1, x2, args, tol, rtol, maxiter, full_output)
+        else:
+           raise ValueError("x0, x1, x2 arrays must be of the same size")
+
+    funcalls = 0
+
+    # Ensure inputs are numpy arrays and converted to complex.
+    p0 = np.asarray(x0, dtype=complex)[()] * 1.0
+    p1 = np.asarray(x1, dtype=complex)[()] * 1.0
+    p2 = np.asarray(x2, dtype=complex)[()] * 1.0
+
```

Comment:
@mdhaber note that `x0` etc. are already cast to complex. If by "require" you mean throw an error/warning if inputs aren't already complex, I could do that.

Reply:
Ah, that might not be necessary. We would just need to make sure it gets cast to the appropriate precision when we're working with arrays. No need to do that right now. 

I agree with Jake that we should start with the scalar version written as a private function but exposed through `root_scalar`.  Please refactor like that and we can start reviewing the code in detail.